### DISCUSSION TITLE
What does "Engine_Interface_FDTD::CalcVoltageIntegral: Error, only a 1D/line integration is allowed" mean?

### DISCUSSION THREAD

**Discussion by XDAChen** [thliebig/openEMS-Project#81] [Q&A Usage]
Hi, I am trying to run a simulation with an imported, self-generated bent surface with a source. 
Here is the image:

![Capture](https://github.com/thliebig/openEMS-Project/assets/47087947/5d5a3702-b5f5-4b25-b0af-81d6c5ba76dc)
 
The shape is imported from an .stl file using the following code:
`%% add geometry
blender_model = '/test.stl';
CSX = AddMetal( CSX, 'copper' );
CSX = ImportSTL(CSX, 'copper',0, [currDir blender_model],'Transform',{'Scale', 1e4});
`
but when I execute RunOpenEMS(), I am getting a repeat error:
`Engine_Interface_FDTD::CalcVoltageIntegral: Error, only a 1D/line integration is allowed 
Engine_Interface_FDTD::CalcVoltageIntegral: Error, only a 1D/line integration is allowed `

anyone knows what that means or what could go wrong with my simulation?

**Comment by thliebig** [comment]
This will not work like this, but I'm not even sure the Error messages are even related to your metal objects? 

Firstly, just a surface cannot be imported. Every material object has to have a volume... The object looks more like a surface only?
What openEMS currently is lacking is a good support for metal surfaces (zero volume objects).
Ideally some day we would need an algorithm that maps surfaces onto FDTD edges and planes. But currently we only have such a mapping for curves (AddCurve)

The error message relates to voltage integrals like in a port? I do not see how that relates to your STL-metal?

**Reply by XDAChen** [reply]
thank you @thliebig  for the explanation. It turns out the error message is indeed related to the port. I made a dummy box that is not connected to my shape and the simulation ran without the error showing.

In terms of surface simulation. This is part of my project that I am focusing on surface deformation. But the .stl file is generated using MATLAB, I added some thickness to the surface (a bit hard to see) in terms of point clouds so to give it some volume and then convert to .stl. But MATLAB is difficult in converting point clouds into .stl and currently the easy solution is to get the volume's outer surface and write it into .stl, and that is what you are seeing in the post. 

Right now I am still investigating 1)if there will be pitfalls between writing a volume's outer surface into .stl and writing the volume directly in .stl. .. Particularly I need to make sure the former approach, when converting into .stl would give a volume. If not I will write a Matlab script myself to convert point-could directly into a .stl file.  So far when I use the dummy port the simulation run without complaint 2) The port shape(or placement) in the post apparently is giving an error, I need to figure out why.

I will keep updating in this post.

**Comment by biergaizi** [comment]
If the openEMS output is flooded with the following error message:

```
Engine_Interface_FDTD::CalcVoltageIntegral: Error, only a 1D/line integration is allowed
Engine_Interface_FDTD::CalcVoltageIntegral: Error, only a 1D/line integration is allowed
```

It means that openEMS is not able to calculate the voltage at a port because the port is located at on ill-defined position. This happens if the start and stop coordinates are different (i.e. not a 1D port), but the size is smaller than a single mesh cell (i.e. when the port is built from its start coordinate to its stop coordinate on each axis, it does not overlap with at least two mesh lines).

For example, the following port is functional because it’s strictly a 1D port:

```
port[0] = fdtd.AddLumpedPort(1, z0, [any_x, any_y, -8], [any_x, any_y, -8], 'y', excite=1)
port[1] = fdtd.AddLumpedPort(2, z0, [any_x, any_y, -8], [any_x, any_y, -8], 'y', excite=0)
```

The following port is also functional because the 2D port passes (overlaps) with at least two mesh lines when it’s built from Z = -8 to Z = 8:

```
port[0] = fdtd.AddLumpedPort(1, z0, [any_x, any_y, -8], [any_x, any_y, 8], 'z', excite=1)
port[1] = fdtd.AddLumpedPort(2, z0, [any_x, any_y, -8], [any_x, any_y, 8], 'z', excite=0)
```

The following port is also functional, because although there is no mesh line at the stop position Z = 8.1, but the 2D port has already crossed at least one mesh cell (two mesh lines) when it’s built from Z = -8 to its stop coordinate:

```
port[0] = fdtd.AddLumpedPort(1, z0, [any_x, any_y, -8], [any_x, any_y, 8.1], 'z', excite=1)
port[1] = fdtd.AddLumpedPort(2, z0, [any_x, any_y, -8], [any_x, any_y, 8.1], 'z', excite=0)
```

But the following port is not functional, because the 2D port does not cross a single mesh cell (two mesh lines) when it’s built from Z = -8 to Z = -7.9. Although there’s a mesh line at Z = -8, there is no mesh line between Z = -8 and Z = -7.9.

    port[0] = fdtd.AddLumpedPort(1, z0, [any_x, any_y, -8], [any_x, any_y, -7.9], 'z', excite=1)
    port[1] = fdtd.AddLumpedPort(2, z0, [any_x, any_y, -8], [any_x, any_y, -7.9], 'z', excite=0)

## Conclusion

On each axis, a port must either be one-dimensional and aligned to a mesh line, or two-dimensional and crosses (or overlaps) with two mesh lines (a single mesh cell). Otherwise it can't be simulated.

Furthermore, a port's excitation direction must be two-dimensional. If the port excites the Z direction, it must have a length on the Z axis, satisfying the two constraints mentioned.

**Reply by is-it-ayush** [reply]
Thanks! This makes sense. In simple words, one can fix this by making sure their mesh is smaller than port dimension in that axis where this failure is occuring.

**Reply by biergaizi** [reply]
@is-it-ayush There's currently a massive project to document openEMS at https://github.com/thliebig/openEMS-Project/pull/417. You may find the following draft documentation helpful. Hopefully they'll soon become official in early 2026.

* https://openems--417.org.readthedocs.build/en/417/concepts/index.html
* https://fasterems.github.io/openEMS-Project/python/openEMS/Tutorials/First_Lessons.html

**Comment by VolkerMuehlhaus** [comment]
I'm confused.

You have a single conductor sheet (curved), so what port are you trying to apply there? Waveguide port makes no sense because the zero thickness sheet isn't a hollow waveguide. Other ports also don't make sense, because they need to apply a signal between two conductors. So what are you tring to do here?

That said, the mesh density in your screenshot is much to coarse to sample the curvature of your geometry.

And regarding the STL file import issues ... if you generate the model yourself anyway, wouldn't it be easier to avoid the STL import, and create the entire model using openEMS geometry commands like [AddLinPoly()](https://docs.openems.de/python/CSXCAD/CSProperties/CSProperties.html#CSXCAD.CSProperties.CSProperties.AddLinPoly)?
